tgui/packages/tgui/interfaces/Manufacturer/index.tsx

```diff
@@ -0,0 +1,379 @@
+      - Eject button
+      - Priority swap buttons (swap priority of two materials for fabricator use)
+      - Speed control buttons
+      - Repeat toggle button
+      - Ore buying which contains ores by rockbox
+
+      Each rockbox element contains:
+        - List of ores available for purchase
+
+        Each ore available for purchase contains:
+          - Button with ore icon, name, amount, and cost
+*/
+
+const BlueprintTooltip = (props) => {
+  let { blueprint } = props;
+  let requirements = [];
+  for (let i in blueprint.item_names) {
+    requirements.push(<>{blueprint.item_amounts[i]/10} {blueprint.item_names[i]}<br /></>);
+  }
+  return (
+    <>
+      Resource Costs:<br />
+      {requirements} <br />
+      {formatTime(blueprint.time, "Instananeous")}
+    </>
+  );
+};
+
+const BlueprintButton = (props, context) => {
+  const { act } = useBackend(context);
+  const { name, blueprintData } = props;
+
+  return (
+    <Stack width={15} inline backgroundColor={backgroundPop} mx={0.575} mb={0} p={0.5}>
+      <Stack.Item>
+        <Button
+          ellipsis
+          width={12.2}
+          height={5.3}
+          p={0}
+          onClick={() => act("product", { "blueprint_ref": blueprintData.ref })}
+        >
+          <Stack>
+            <Stack.Item
+              ml={0}
+              pt={0}
+              mt={0}
+              backgroundColor={backgroundPop}
+            >
+              <Image pixelated src={blueprintData.img} width={5} />
+            </Stack.Item>
+            <Stack.Item>
+              <Stack
+                vertical
+                fill
+                style={{ "align-items": "center" }}
+              >
+                <Stack.Item
+                  width={6}
+                  textAlign="center"
+                  style={{ "white-space": "normal" }}
+                >
+                  {truncate(name, 30)}
+                </Stack.Item>
+              </Stack>
+            </Stack.Item>
+          </Stack>
+        </Button>
+      </Stack.Item>
+
+      <Stack.Item ml={0.5}>
+        <Stack vertical p={0} m={0}>
+          <Stack.Item pb={0.5}>
+            <Button align="center" width={2} height={2.5} icon="gear" pt={0.9} />
+          </Stack.Item>
+          <Stack.Item p={0} m={0}>
+            <Tooltip content={<BlueprintTooltip blueprint={blueprintData} />}>
+              <Button align="center" width={2} height={2.5} icon="question" pt={0.9} />
+            </Tooltip>
+          </Stack.Item>
+        </Stack>
+      </Stack.Item>
+
+    </Stack>
+  );
+};
+
+const MaterialRow = (props, context) => {
+  let { act } = useBackend(context);
+  let { resource, resourceAmt } = props;
+
+  return (
+    <TableRow collapsing>
+      <TableCell collapsing>
+        <Button icon="eject" onClick={() => act("material_eject", { "material": resource })} />
+      </TableCell>
+      <TableCell>
+        <Button icon="add" onClick={() => act("material_swap", { "resource": resource })} />
```

Comment:
`add` button has an action of `material_swap`? 🤔

Reply:
icon is to be changed once i get to implementing material swaps, just needed a placeholder while i worked on the rest of this GUI to find something that feels right.

Reply:
and yeah plus icon is not staying i fully agree. a lot of the ideas of things to change are in my head right now and its difficult to collect that all and say "these are the parts i know need work in these ways" but hopefully after work tomorrow and after making the code saner i can do that